File: BE/indive/src/main/java/com/ssafy/indive/global/blockchain/InDive.java
package com.ssafy.indive.global.blockchain;

import io.reactivex.Flowable;
import io.reactivex.functions.Function;
import java.math.BigInteger;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.Callable;
import org.web3j.abi.EventEncoder;
import org.web3j.abi.TypeReference;
import org.web3j.abi.datatypes.Address;
import org.web3j.abi.datatypes.Bool;
import org.web3j.abi.datatypes.Event;
import org.web3j.abi.datatypes.Type;
import org.web3j.abi.datatypes.Utf8String;
import org.web3j.abi.datatypes.generated.Uint256;
import org.web3j.crypto.Credentials;
import org.web3j.protocol.Web3j;
import org.web3j.protocol.core.DefaultBlockParameter;
import org.web3j.protocol.core.RemoteFunctionCall;
import org.web3j.protocol.core.methods.request.EthFilter;
import org.web3j.protocol.core.methods.response.BaseEventResponse;
import org.web3j.protocol.core.methods.response.Log;
import org.web3j.protocol.core.methods.response.TransactionReceipt;
import org.web3j.tuples.generated.Tuple2;
import org.web3j.tuples.generated.Tuple6;
import org.web3j.tx.Contract;
import org.web3j.tx.TransactionManager;
import org.web3j.tx.gas.ContractGasProvider;

/**
 * <p>Auto generated code.
 * <p><strong>Do not modify!</strong>
 * <p>Please use the <a href="https://docs.web3j.io/command_line.html">web3j command line tools</a>,
 * or the org.web3j.codegen.SolidityFunctionWrapperGenerator in the 
 * <a href="https://github.com/web3j/web3j/tree/master/codegen">codegen module</a> to update.
 *
 * <p>Generated with web3j version 1.4.1.
 */
@SuppressWarnings("rawtypes")
public class InDive extends Contract {
    public static final String BINARY = "Bin file was not provided";

    public static final String FUNC_DONATE = "donate";

    public static final String FUNC_DONATIONHISTORY = "donationHistory";

    public static final String FUNC_DONATORLIST = "donatorList";

    public static final String FUNC_GETDONATIONHISTORYLIST = "getDonationHistoryList";

    public static final String FUNC_GETDONATORLIST = "getDonatorList";

    public static final String FUNC_ISDONATED = "isDonated";

    public static final String FUNC_SETTOKENCONTRACT = "setTokenContract";

    public static final Event DONATIONEVENT_EVENT = new Event("DonationEvent", 
            Arrays.<TypeReference<?>>asList(new TypeReference<Address>(true) {}, new TypeReference<Address>(true) {}, new TypeReference<Uint256>() {}, new TypeReference<Utf8String>() {}, new TypeReference<Uint256>() {}));
    ;

    @Deprecated
    protected InDive(String contractAddress, Web3j web3j, Credentials credentials, BigInteger gasPrice, BigInteger gasLimit) {
        super(BINARY, contractAddress, web3j, credentials, gasPrice, gasLimit);
    }

    protected InDive(String contractAddress, Web3j web3j, Credentials credentials, ContractGasProvider contractGasProvider) {
        super(BINARY, contractAddress, web3j, credentials, contractGasProvider);
    }

    @Deprecated
    protected InDive(String contractAddress, Web3j web3j, TransactionManager transactionManager, BigInteger gasPrice, BigInteger gasLimit) {
        super(BINARY, contractAddress, web3j, transactionManager, gasPrice, gasLimit);
    }

    protected InDive(String contractAddress, Web3j web3j, TransactionManager transactionManager, ContractGasProvider contractGasProvider) {
        super(BINARY, contractAddress, web3j, transactionManager, contractGasProvider);
    }

    public List<DonationEventEventResponse> getDonationEventEvents(TransactionReceipt transactionReceipt) {
        List<Contract.EventValuesWithLog> valueList = extractEventParametersWithLog(DONATIONEVENT_EVENT, transactionReceipt);
        ArrayList<DonationEventEventResponse> responses = new ArrayList<DonationEventEventResponse>(valueList.size());
        for (Contract.EventValuesWithLog eventValues : valueList) {
            DonationEventEventResponse typedResponse = new DonationEventEventResponse();
            typedResponse.log = eventValues.getLog();
            typedResponse.artist = (String) eventValues.getIndexedValues().get(0).getValue();
            typedResponse.donator = (String) eventValues.getIndexedValues().get(1).getValue();
            typedResponse.value = (BigInteger) eventValues.getNonIndexedValues().get(0).getValue();
            typedResponse.message = (String) eventValues.getNonIndexedValues().get(1).getValue();
            typedResponse.time = (BigInteger) eventValues.getNonIndexedValues().get(2).getValue();
            responses.add(typedResponse);
        }
        return responses;
    }

    public Flowable<DonationEventEventResponse> donationEventEventFlowable(EthFilter filter) {
        return web3j.ethLogFlowable(filter).map(new Function<Log, DonationEventEventResponse>() {
            @Override
            public DonationEventEventResponse apply(Log log) {
                Contract.EventValuesWithLog eventValues = extractEventParametersWithLog(DONATIONEVENT_EVENT, log);
                DonationEventEventResponse typedResponse = new DonationEventEventResponse();
                typedResponse.log = log;
                typedResponse.artist = (String) eventValues.getIndexedValues().get(0).getValue();
                typedResponse.donator = (String) eventValues.getIndexedValues().get(1).getValue();
                typedResponse.value = (BigInteger) eventValues.getNonIndexedValues().get(0).getValue();
                typedResponse.message = (String) eventValues.getNonIndexedValues().get(1).getValue();
                typedResponse.time = (BigInteger) eventValues.getNonIndexedValues().get(2).getValue();
                return typedResponse;
            }
        });
    }

    public Flowable<DonationEventEventResponse> donationEventEventFlowable(DefaultBlockParameter startBlock, DefaultBlockParameter endBlock) {
        EthFilter filter = new EthFilter(startBlock, endBlock, getContractAddress());
        filter.addSingleTopic(EventEncoder.encode(DONATIONEVENT_EVENT));
        return donationEventEventFlowable(filter);
    }

    public RemoteFunctionCall<TransactionReceipt> donate(String _to, BigInteger _value, String _message, BigInteger _time) {
        final org.web3j.abi.datatypes.Function function = new org.web3j.abi.datatypes.Function(
                FUNC_DONATE, 
                Arrays.<Type>asList(new org.web3j.abi.datatypes.Address(160, _to), 
                new org.web3j.abi.datatypes.generated.Uint256(_value), 
                new org.web3j.abi.datatypes.Utf8String(_message), 
                new org.web3j.abi.datatypes.generated.Uint256(_time)), 
                Collections.<TypeReference<?>>emptyList());
        return executeRemoteCallTransaction(function);
    }

    public RemoteFunctionCall<Tuple6<BigInteger, String, String, BigInteger, String, BigInteger>> donationHistory(String param0, BigInteger param1) {
        final org.web3j.abi.datatypes.Function function = new org.web3j.abi.datatypes.Function(FUNC_DONATIONHISTORY, 
                Arrays.<Type>asList(new org.web3j.abi.datatypes.Address(160, param0), 
                new org.web3j.abi.datatypes.generated.Uint256(param1)), 
                Arrays.<TypeReference<?>>asList(new TypeReference<Uint256>() {}, new TypeReference<Address>() {}, new TypeReference<Utf8String>() {}, new TypeReference<Uint256>() {}, new TypeReference<Utf8String>() {}, new TypeReference<Uint256>() {}));
        return new RemoteFunctionCall<Tuple6<BigInteger, String, String, BigInteger, String, BigInteger>>(function,
                new Callable<Tuple6<BigInteger, String, String, BigInteger, String, BigInteger>>() {
                    @Override
                    public Tuple6<BigInteger, String, String, BigInteger, String, BigInteger> call() throws Exception {
                        List<Type> results = executeCallMultipleValueReturn(function);
                        return new Tuple6<BigInteger, String, String, BigInteger, String, BigInteger>(
                                (BigInteger) results.get(0).getValue(), 
                                (String) results.get(1).getValue(), 
                                (String) results.get(2).getValue(), 
                                (BigInteger) results.get(3).getValue(), 
                                (String) results.get(4).getValue(), 
                                (BigInteger) results.get(5).getValue());
                    }
                });
    }

    public RemoteFunctionCall<Tuple2<String, BigInteger>> donatorList(String param0, BigInteger param1) {
        final org.web3j.abi.datatypes.Function function = new org.web3j.abi.datatypes.Function(FUNC_DONATORLIST, 
                Arrays.<Type>asList(new org.web3j.abi.datatypes.Address(160, param0), 
                new org.web3j.abi.datatypes.generated.Uint256(param1)), 
                Arrays.<TypeReference<?>>asList(new TypeReference<Address>() {}, new TypeReference<Uint256>() {}));
        return new RemoteFunctionCall<Tuple2<String, BigInteger>>(function,
                new Callable<Tuple2<String, BigInteger>>() {
                    @Override
                    public Tuple2<String, BigInteger> call() throws Exception {
                        List<Type> results = executeCallMultipleValueReturn(function);
                        return new Tuple2<String, BigInteger>(
                                (String) results.get(0).getValue(), 
                                (BigInteger) results.get(1).getValue());
                    }
                });
    }

    public RemoteFunctionCall<String> getDonationHistoryList(String artist) {
        final org.web3j.abi.datatypes.Function function = new org.web3j.abi.datatypes.Function(FUNC_GETDONATIONHISTORYLIST, 
                Arrays.<Type>asList(new org.web3j.abi.datatypes.Address(160, artist)), 
                Arrays.<TypeReference<?>>asList(new TypeReference<Utf8String>() {}));
        return executeRemoteCallSingleValueReturn(function, String.class);
    }

    public RemoteFunctionCall<String> getDonatorList(String artist) {
        final org.web3j.abi.datatypes.Function function = new org.web3j.abi.datatypes.Function(FUNC_GETDONATORLIST, 
                Arrays.<Type>asList(new org.web3j.abi.datatypes.Address(160, artist)), 
                Arrays.<TypeReference<?>>asList(new TypeReference<Utf8String>() {}));
        return executeRemoteCallSingleValueReturn(function, String.class);
    }

    public RemoteFunctionCall<Boolean> isDonated(String param0, String param1) {
        final org.web3j.abi.datatypes.Function function = new org.web3j.abi.datatypes.Function(FUNC_ISDONATED, 
                Arrays.<Type>asList(new org.web3j.abi.datatypes.Address(160, param0), 
                new org.web3j.abi.datatypes.Address(160, param1)), 
                Arrays.<TypeReference<?>>asList(new TypeReference<Bool>() {}));
        return executeRemoteCallSingleValueReturn(function, Boolean.class);
    }

    public RemoteFunctionCall<TransactionReceipt> setTokenContract(String contractAddress) {
        final org.web3j.abi.datatypes.Function function = new org.web3j.abi.datatypes.Function(
                FUNC_SETTOKENCONTRACT, 
                Arrays.<Type>asList(new org.web3j.abi.datatypes.Address(160, contractAddress)), 
                Collections.<TypeReference<?>>emptyList());
        return executeRemoteCallTransaction(function);
    }

    @Deprecated
    public static InDive load(String contractAddress, Web3j web3j, Credentials credentials, BigInteger gasPrice, BigInteger gasLimit) {
        return new InDive(contractAddress, web3j, credentials, gasPrice, gasLimit);
    }

    @Deprecated
    public static InDive load(String contractAddress, Web3j web3j, TransactionManager transactionManager, BigInteger gasPrice, BigInteger gasLimit) {
        return new InDive(contractAddress, web3j, transactionManager, gasPrice, gasLimit);
    }

    public static InDive load(String contractAddress, Web3j web3j, Credentials credentials, ContractGasProvider contractGasProvider) {
        return new InDive(contractAddress, web3j, credentials, contractGasProvider);
    }

    public static InDive load(String contractAddress, Web3j web3j, TransactionManager transactionManager, ContractGasProvider contractGasProvider) {
        return new InDive(contractAddress, web3j, transactionManager, contractGasProvider);
    }

    public static class DonationEventEventResponse extends BaseEventResponse {
        public String artist;

        public String donator;

        public BigInteger value;

        public String message;

        public BigInteger time;
    }
}

File: AOS/InDive/app/src/main/java/com/ssafy/indive/blockchain/InDive.kt
package com.ssafy.indive.blockchain

import org.web3j.protocol.Web3j
import com.ssafy.indive.blockchain.InDive
import org.web3j.tx.gas.ContractGasProvider
import org.web3j.tx.TransactionManager
import org.web3j.protocol.core.RemoteFunctionCall
import org.web3j.protocol.core.methods.response.TransactionReceipt
import org.web3j.abi.datatypes.generated.Uint256
import com.ssafy.indive.blockchain.InDive.DonationHistoryEventResponse
import org.web3j.tx.Contract.EventValuesWithLog
import io.reactivex.Flowable
import io.reactivex.functions.Function
import org.web3j.protocol.core.DefaultBlockParameter
import org.web3j.abi.EventEncoder
import org.web3j.abi.TypeReference
import org.web3j.abi.datatypes.*
import org.web3j.crypto.Credentials
import org.web3j.protocol.core.methods.request.EthFilter
import org.web3j.tuples.generated.Tuple2
import kotlin.Throws
import org.web3j.protocol.core.methods.response.BaseEventResponse
import org.web3j.protocol.core.methods.response.Log
import org.web3j.tx.Contract
import java.math.BigInteger
import java.util.*

/**
 *
 * Auto generated code.
 *
 * **Do not modify!**
 *
 * Please use the [web3j command line tools](https://docs.web3j.io/command_line.html),
 * or the org.web3j.codegen.SolidityFunctionWrapperGenerator in the
 * [codegen module](https://github.com/web3j/web3j/tree/master/codegen) to update.
 *
 *
 * Generated with web3j version 1.4.1.
 */
class InDive : Contract {
    @Deprecated("")
    protected constructor(
        contractAddress: String?,
        web3j: Web3j?,
        credentials: Credentials?,
        gasPrice: BigInteger?,
        gasLimit: BigInteger?
    ) : super(
        BINARY, contractAddress, web3j, credentials, gasPrice, gasLimit
    ) {
    }

    protected constructor(
        contractAddress: String?,
        web3j: Web3j?,
        credentials: Credentials?,
        contractGasProvider: ContractGasProvider?
    ) : super(
        BINARY, contractAddress, web3j, credentials, contractGasProvider
    ) {
    }

    @Deprecated("")
    protected constructor(
        contractAddress: String?,
        web3j: Web3j?,
        transactionManager: TransactionManager?,
        gasPrice: BigInteger?,
        gasLimit: BigInteger?
    ) : super(
        BINARY, contractAddress, web3j, transactionManager, gasPrice, gasLimit
    ) {
    }

    protected constructor(
        contractAddress: String?,
        web3j: Web3j?,
        transactionManager: TransactionManager?,
        contractGasProvider: ContractGasProvider?
    ) : super(
        BINARY, contractAddress, web3j, transactionManager, contractGasProvider
    ) {
    }

    fun donate(
        _to: String?,
        _value: BigInteger?,
        _message: String?
    ): RemoteFunctionCall<TransactionReceipt> {
        val function = Function(
            FUNC_DONATE,
            Arrays.asList<Type<*>>(
                Address(160, _to),
                Uint256(_value),
                Utf8String(_message)
            ), emptyList()
        )
        return executeRemoteCallTransaction(function)
    }

    fun setTokenContract(contractAddress: String?): RemoteFunctionCall<TransactionReceipt> {
        val function = Function(
            FUNC_SETTOKENCONTRACT,
            Arrays.asList<Type<*>>(Address(160, contractAddress)), emptyList()
        )
        return executeRemoteCallTransaction(function)
    }

    fun getDonationHistoryEvents(transactionReceipt: TransactionReceipt?): List<DonationHistoryEventResponse> {
        val valueList = extractEventParametersWithLog(DONATIONHISTORY_EVENT, transactionReceipt)
        val responses = ArrayList<DonationHistoryEventResponse>(valueList.size)
        for (eventValues in valueList) {
            val typedResponse = DonationHistoryEventResponse()
            typedResponse.log = eventValues.log
            typedResponse.artist = eventValues.nonIndexedValues[0].value as String
            typedResponse.donator = eventValues.nonIndexedValues[1].value as String
            typedResponse.value = eventValues.nonIndexedValues[2].value as BigInteger
            typedResponse.message = eventValues.nonIndexedValues[3].value as String
            responses.add(typedResponse)
        }
        return responses
    }

    fun donationHistoryEventFlowable(filter: EthFilter?): Flowable<DonationHistoryEventResponse> {
        return web3j.ethLogFlowable(filter)
            .map(object : Function<Log?, DonationHistoryEventResponse> {
                override fun apply(t: Log): DonationHistoryEventResponse {
                    val eventValues = extractEventParametersWithLog(DONATIONHISTORY_EVENT, t)
                    val typedResponse = DonationHistoryEventResponse()
                    typedResponse.log = t
                    typedResponse.artist = eventValues.nonIndexedValues[0].value as String
                    typedResponse.donator = eventValues.nonIndexedValues[1].value as String
                    typedResponse.value = eventValues.nonIndexedValues[2].value as BigInteger
                    typedResponse.message = eventValues.nonIndexedValues[3].value as String
                    return typedResponse
                }

            })
    }

    fun donationHistoryEventFlowable(
        startBlock: DefaultBlockParameter?,
        endBlock: DefaultBlockParameter?
    ): Flowable<DonationHistoryEventResponse> {
        val filter = EthFilter(startBlock, endBlock, getContractAddress())
        filter.addSingleTopic(EventEncoder.encode(DONATIONHISTORY_EVENT))
        return donationHistoryEventFlowable(filter)
    }

    fun donatorList(
        param0: String?,
        param1: BigInteger?
    ): RemoteFunctionCall<Tuple2<String, BigInteger>> {
        val function = Function(
            FUNC_DONATORLIST,
            Arrays.asList<Type<*>>(
                Address(160, param0),
                Uint256(param1)
            ),
            Arrays.asList<TypeReference<*>>(
                object : TypeReference<Address?>() {},
                object : TypeReference<Uint256?>() {})
        )
        return RemoteFunctionCall(
            function
        ) {
            val results = executeCallMultipleValueReturn(function)
            Tuple2(
                results[0].value as String,
                results[1].value as BigInteger
            )
        }
    }

    fun getDonatorList(artist: String?): RemoteFunctionCall<String> {
        val function = Function(
            FUNC_GETDONATORLIST,
            Arrays.asList<Type<*>>(Address(160, artist)),
            Arrays.asList<TypeReference<*>>(object : TypeReference<Utf8String?>() {})
        )
        return executeRemoteCallSingleValueReturn(function, String::class.java)
    }

    fun isDonated(param0: String?, param1: String?): RemoteFunctionCall<Boolean> {
        val function = Function(
            FUNC_ISDONATED,
            Arrays.asList<Type<*>>(
                Address(160, param0),
                Address(160, param1)
            ),
            Arrays.asList<TypeReference<*>>(object : TypeReference<Bool?>() {})
        )
        return executeRemoteCallSingleValueReturn(function, Boolean::class.java)
    }

    class DonationHistoryEventResponse : BaseEventResponse() {
        var artist: String? = null
        var donator: String? = null
        var value: BigInteger? = null
        var message: String? = null
    }

    companion object {
        const val BINARY = "Bin file was not provided"
        const val FUNC_DONATE = "donate"
        const val FUNC_SETTOKENCONTRACT = "setTokenContract"
        const val FUNC_DONATORLIST = "donatorList"
        const val FUNC_GETDONATORLIST = "getDonatorList"
        const val FUNC_ISDONATED = "isDonated"
        val DONATIONHISTORY_EVENT = Event(
            "DonationHistory",
            Arrays.asList<TypeReference<*>>(
                object : TypeReference<Address?>() {},
                object : TypeReference<Address?>() {},
                object : TypeReference<Uint256?>() {},
                object : TypeReference<Utf8String?>() {})
        )

        @Deprecated("")
        fun load(
            contractAddress: String?,
            web3j: Web3j?,
            credentials: Credentials?,
            gasPrice: BigInteger?,
            gasLimit: BigInteger?
        ): InDive {
            return InDive(contractAddress, web3j, credentials, gasPrice, gasLimit)
        }

        @Deprecated("")
        fun load(
            contractAddress: String?,
            web3j: Web3j?,
            transactionManager: TransactionManager?,
            gasPrice: BigInteger?,
            gasLimit: BigInteger?
        ): InDive {
            return InDive(contractAddress, web3j, transactionManager, gasPrice, gasLimit)
        }

        fun load(
            contractAddress: String?,
            web3j: Web3j?,
            credentials: Credentials?,
            contractGasProvider: ContractGasProvider?
        ): InDive {
            return InDive(contractAddress, web3j, credentials, contractGasProvider)
        }

        fun load(
            contractAddress: String?,
            web3j: Web3j?,
            transactionManager: TransactionManager?,
            contractGasProvider: ContractGasProvider?
        ): InDive {
            return InDive(contractAddress, web3j, transactionManager, contractGasProvider)
        }
    }
}